pkg/gitignore/dir.go

```diff
@@ -0,0 +1,104 @@
+package gitignore
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-git/go-git/plumbing/format/gitignore"
+	"github.com/spf13/afero"
+)
+
+const (
+	commentPrefix   = "#"
+	coreSection     = "core"
+	eol             = "\n"
+	excludesfile    = "excludesfile"
+	gitDir          = ".git"
+	gitignoreFile   = ".gitignore"
+	gitconfigFile   = ".gitconfig"
+	systemFile      = "/etc/gitconfig"
+	infoExcludeFile = gitDir + "/info/exclude"
+)
+
+// ReadPatterns reads gitignore patterns recursively traversing through the directory
+// structure. The result is in the ascending order of priority (last higher).
+func ReadPatterns(fs afero.Fs, path []string) ([]gitignore.Pattern, error) {
+	path = path[:len(path):len(path)] // ensures that append does not use excess capacity
+
+	ps, err := readIgnoreFile(fs, path, infoExcludeFile)
+	if err != nil {
+		return nil, err
+	}
+
+	subps, err := readIgnoreFile(fs, path, gitignoreFile)
+	if err != nil {
+		return nil, err
+	}
+	ps = append(ps, subps...)
+
+	var fis []os.FileInfo
+	fis, err = readDir(fs, filepath.Join(path...))
+	if err != nil {
+		return nil, err
+	}
+
+	for _, fi := range fis {
+		if fi.IsDir() && fi.Name() != gitDir {
+			// check if the directory is ignored by previous patterns
+			if gitignore.NewMatcher(ps).Match(append(path, fi.Name()), true) {
+				continue
+			}
+
+			var subps []gitignore.Pattern
+			subps, err := ReadPatterns(fs, append(path, fi.Name()))
+			if err != nil {
+				return nil, err
+			}
+
+			if len(subps) > 0 {
+				ps = append(ps, subps...)
+			}
+		}
+	}
+
+	return ps, nil
+}
+
+// readIgnoreFile reads a specific .gitignore file
+func readIgnoreFile(fs afero.Fs, path []string, ignoreFile string) (ps []gitignore.Pattern, err error) {
+	path = path[:len(path):len(path)]
+
+	f, err := fs.Open(filepath.Join(append(path, ignoreFile)...)) // this might cause an issue
+	if err != nil {
+		// .gitignore does not exist along this path
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer f.Close()
+
+	data, err := io.ReadAll(f)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range strings.Split(string(data), eol) {
+		if !strings.HasPrefix(s, commentPrefix) && len(strings.TrimSpace(s)) > 0 {
+			ps = append(ps, gitignore.ParsePattern(s, path))
+		}
+	}
+
+	return ps, nil
+}
+
+func readDir(fs afero.Fs, dir string) ([]os.FileInfo, error) {
+	d, err := fs.Open(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.Readdir(0)
```

Comment:
what's 0? what does it mean?

Reply:
it's a count that you can specify in the afero implementation. Reads all if set to zero.

Reply:
https://github.com/artmoskvin/hide/blob/8a2c595dd54dbe586de18a04cdb5e1ed63a5f992/vendor/github.com/spf13/afero/mem/file.go#L160